internal/Yesod/Session/Memcache/Storage.hs

```diff
@@ -0,0 +1,94 @@
+module Yesod.Session.Memcache.Storage
+  ( memcacheStorage
+  , SessionPersistence (..)
+  ) where
+
+import Internal.Prelude
+
+import Control.Monad.Reader (MonadReader (ask))
+import Database.Memcache.Client qualified as Memcache
+import Database.Memcache.Types qualified as Memcache
+import Session.Key
+import Session.Timing.Options (TimingOptions (timeout))
+import Session.Timing.Time (Time (..))
+import Time (UTCTime)
+import Yesod.Core (SessionMap)
+import Yesod.Session.Memcache.Expiration
+  ( MemcacheExpiration
+  , getCacheExpiration
+  , noExpiration
+  )
+import Yesod.Session.Options (Options (clock, timing))
+import Yesod.Session.SessionType
+import Yesod.Session.Storage.Exceptions
+import Yesod.Session.Storage.Operation
+
+-- | Mapping between 'Session' and Memcache representation.
+data SessionPersistence = SessionPersistence
+  { databaseKey :: SessionKey -> Memcache.Key
+  , toDatabase :: (SessionMap, Time UTCTime) -> Memcache.Value
+  , fromDatabase :: Memcache.Value -> (SessionMap, Time UTCTime)
```

Comment:
I wonder if we should move away from the name "database" here, as more than just databases can (and now do) at as session-persistence.

Reply:
Well, reasonable people could disagree. You could get pedantic and distinguish "key value store" from "database" and say no, but that's not the point I was making here.

What' I'm saying is, what about not-even-memcached cases? Can't this abstraction work if you "persist" the data in memory, or a file, or via HTTP calls to another service? Nothing here says it needs to be a database right?

Reply:
Fair enough. `toMemcacheValue` and `fromMemcacheValue` seem clearer.

Reply:
Wait, what?

Oh I'm sorry, I was a little turned around.

I assumed the `SessionPersistence` type was abstract over whatever the backend may be, something like,

```hs
data SessionPersistence k v = SessionPersistence
  { databaseKey :: SessionKey -> k
  , toDatabase :: (SessionMap, Time UTCTime) -> v
  , fromDatabase :: v -> (SessionMap, TimeUTCTime)
  }
```

And this was just the specific `type MemcacheSessionPersistence = SessionPersistence Memcache.Key Memcache.Value`. But I see now this record is entirely `Memcache`-specific.

OK, call it what you like. And the "memcache is kind of a database" naming take is totally fine to lean on for it. I was just surprised to not see a polymorphic record (in which case my naming question would make more sense) being used here.